hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3AFileSystem.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.s3a;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.regex.Pattern;
+
+/**
+ * Unit tests for {@link S3AFileSystem}.
+ */
+public class TestS3AFileSystem extends Assert {
+  final File TEMP_DIR = new File("target/build/test/TestS3AFileSystem");
+  final String longStr = // 1024 char
+    "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key__very_long_s3_key__very_long_s3_key__very_long_s3_key__" +
+      "very_long_s3_key";
+  final String longStrTruncated = "very_long_s3_key__very_long_s3_key__";
+
+  @Rule
+  public Timeout testTimeout = new Timeout(30 * 1000);
+
+  @Before
+  public void init() throws IOException {
+    Files.createDirectories(TEMP_DIR.toPath());
```

Comment:
use `S3ATestUtils` code here for handling of parallel test cases and ease of future maintenance
```
Configuration conf = new Configuration(false)
S3ATestUtils.prepareTestConfiguration(conf)
conf.get(BUFFER_DIR)
```